### ISSUE TITLE
Step does not support multiple destinations

### ISSUE THREAD

**Issue by dagio** [bitrise-io/steps-xcode-test#51]
The step assume only one device is used while Xcode allow to use multiple destinations

**Comment by viktorbenei** [comment]
You can add as many Xcode Test steps as you want to. Isn't that an option?

**Comment by dagio** [comment]
Yes but I believe it's faster using multiple destinations in a single step

**Comment by viktorbenei** [comment]
Did you try that? You can copy paste the `xcodebuild` command from this step's log output, and use it in a Script step.

This step should be quite efficient, the overhead of adding a second one vs adding multiple device configs should be minimal (the right simulator have to be booted and the tests have to be performed, you can't avoid that with either solution, and that's what takes the most time).

**Comment by viktorbenei** [comment]
I'm just curious if it would worth it at all - if you can run a couple of tests (multiple Xcode Test step vs Script step) and report the times that would help a lot!

P.S.: just don't do it this weekend, as the performance of the Mac OS stacks can be significantly worse than usual, due to the ongoing infra maintenance ( http://status.bitrise.io/incidents/4swscfprwz0j )

**Comment by dagio** [comment]
Yes that's what I'm actually doing using the script step : 

```
#!/bin/bash
set -ev

xcodebuild \
  -workspace $BITRISE_WORKSPACE_PATH \
  -scheme $BITRISE_SCHEME \
  -sdk iphonesimulator \
  -destination 'platform=iOS Simulator,name=iPhone 4s,OS=9.3' \
  -destination 'platform=iOS Simulator,name=iPhone 6s,OS=9.3' \
  -destination 'platform=iOS Simulator,name=iPhone 6s Plus,OS=9.3' \
 test | xcpretty --no-color -r html --output "${BITRISE_DEPLOY_DIR}/Tests/test-results.html"

exit ${PIPESTATUS[0]}
```

I would have love to make a PR to add multidestinations but I'm not fluent in Go. For now this improvement is not critical for me as the script does the job.

Cheers

**Comment by viktorbenei** [comment]
Thanks @dagio for the script, I know it can be done like this - the question is, is this really faster? Did you measure your script vs doing it with 3 Xcode Test steps?

**Comment by dagio** [comment]
I'll make a little benchmark between few options : 
- fastlane (scan)
- xctool
- xcodebuild w/ multi destinations
- multi xcodebuild w/ single destination

**Comment by viktorbenei** [comment]
Thanks! Please keep in mind that this weekend we perform infra maintenance, which might affect the results. Should be finished later today.

**Comment by kubajakowski** [comment]
`the question is, is this really faster? Did you measure your script vs doing it with 3 Xcode Test steps?`
Of course it is A LOT faster. Consider having end to end UI test suite which can take about 20 minutes to run which you want to run on, lets say, 6 devices. Having 5 test steps means 5x build and run tests one after another. This means over 2 hours of testing. When tested in parallel - it would take 30 minutes max (including git clone, build and so on).

**Comment by viktorbenei** [comment]
@kubajakowski it certainly was not **a LOT** at the time when the issue was created. In 2016 Xcode did not support running tests in parallel. You could specify more than one item but those did run sequentially, one after another (AFAIK).

**Comment by dagio** [comment]
I think now the way to go from Xcode 8 is to use `xcodebuild build-for-testing` then after use `test-without-building` for each devices.

More details : http://shashikantjagtap.net/xcodebuild-deploy-ios-app-command-line/

Note : It's no longer needed from my side so unless someone is interested in that feature, I'm fine if you decide to close it ;-)

**Comment by viktorbenei** [comment]
Thanks @dagio !

I'll leave this issue open here for now, as a reminder, although we did perform some tests with `test-without-building` and found quite a few reliability issues which did not happen with the standard "one test at a time" approach.

We'll of course keep an eye on and keep testing new Xcode versions' parallel testing features.

**Comment by stevelandeyasana** [comment]
Anecdotally, our app's UI tests run in about 60 minutes in serial, and about 30 minutes in parallel.

**Comment by viktorbenei** [comment]
@stevelandeyasana we're mainly concerned about:

> although we did perform some tests with test-without-building and found quite a few reliability issues which did not happen with the standard "one test at a time" approach.

I'll close this ticket here as we moved to a centralized feature request platform since; feel free to create a feature request at: https://discuss.bitrise.io/c/feature-request

That should help with collecting details, discussion and tracking popularity/priority! ;)

**Comment by eggheadgames** [comment]
Xcode 10 will support multiple simulators running at once, as auto-calculated by available memory etc. Early testing on one of our machines suggests that this will provide a significant performance improvement.

**Comment by viktorbenei** [comment]
We'll definitely check & prepare for Xcode 10, but AFAIK it does not require you specifying multiple devices, it can just auto split test suits and run them in parallel using the same Simulator / iOS version. Correct me if I'm wrong @eggheadgames
